src/base/bittorrent/torrenthandle.h

```diff
@@ -255,6 +255,9 @@ namespace BitTorrent
         bool isCompleted() const;
         bool isActive() const;
         bool isInactive() const;
+        bool isStalled() const;
```

Comment:
IMO, TorrentHandle already has bloated interface, and you inflate it even more... why don't use `TorrentHandle::state()` in TorrentFilter?

Reply:
@glassez 
In that case, the code in `TorrentFilter::matchState` will still look somewhat ok:

```cpp
    case Stalled:
        return (torrent->state() ==  BitTorrent::TorrentState::StalledUploading)
                || (torrent->state() ==  BitTorrent::TorrentState::StalledDownloading);
    case StalledUploading:
        return torrent->state() ==  BitTorrent::TorrentState::StalledUploading;
    case StalledDownloading:
        return torrent->state() ==  BitTorrent::TorrentState::StalledDownloading;
```

But consider the case of `StatusFilterWidget::updateTorrentNumbers` (this is what the relevant section looks like currently):

```cpp
        if (torrent->isStalled())
            ++nbStalled;
        if (torrent->isStalledUploading())
            ++nbStalledUploading;
        if (torrent->isStalledDownloading())
            ++nbStalledDownloading;
```

Putting the same logical expressions inside the `if()`s would look very bad and writing helper functions in that source file to return those results seems out of place. Wouldn't it violate the single responsibility principle?

Reply:
>But consider the case of StatusFilterWidget::updateTorrentNumbers (this is what the relevant section looks like currently)
Putting the same logical expressions inside the if()s would look very bad and writing helper functions in that source file to return those results seems out of place.

Just change your mind and look at it from the other side. Num of "Stalled" torrents is just sum of "Stalled Downloading" and "Stalled Uploading" so why not calculate it after loop? Even more, since it's really not unique value why it should be provided here at all? Let client code to calculate it if it's needed for.

Reply:
>Wouldn't it violate the single responsibility principle?

Be careful not to overdo this principle, there are others as well.
Generally there is no sense to have in class interface something that can be trivially calculated from other (basic) public fields.

Reply:
@glassez I see your points. Fixed.